From 19e7cac24fc49cf5aea00dea39ca2ddbfbb90226 Mon Sep 17 00:00:00 2001
From: Wu Zhangjin <wuzhangjin@gmail.com>
Date: Thu, 6 Aug 2009 21:04:44 +0800
Subject: [PATCH] kft: add a high-level tool for tracing kernel function
 automatically

A new tool named autokft.sh is added in scripts/, and the relative
documentation is added to Documentation/kft_kickstart.txt

Signed-off-by: Wu Zhangjin <wuzhangjin@gmail.com>
---
 Documentation/kft_kickstart.txt | 19 ++++++++-
 scripts/autokft.sh              | 86 +++++++++++++++++++++++++++++++++++++++++
 2 files changed, 104 insertions(+), 1 deletion(-)
 create mode 100755 scripts/autokft.sh

diff --git a/Documentation/kft_kickstart.txt b/Documentation/kft_kickstart.txt
index d692007..07d3da7 100644
--- a/Documentation/kft_kickstart.txt
+++ b/Documentation/kft_kickstart.txt
@@ -122,6 +122,23 @@ chmod a+x scripts/addr2sym
 chmod a+x scripts/kd
 ./scripts/kd -c -l -i log.sym
 
+2.3 use it via the autokft.sh
+
+there is a new tool named scripts/autokft.sh which can help you simplifing the
+procedure of tracing an indicated function.
+
+here are several steps to use it:
+
+* compile your kernel with KFT enabled
+* install this new kernel and reboot into it
+* enter into the directory of scripts/
+  or
+  copy the tools scripts/{kd,addr2sym,sym2addr,autokft.sh} to a directory,and enter into it
+* trace and indicated function, here we use sys_write as an example
+  $ ./autokft.sh sys_write ./   # this will be triggered by ourselves
+  or
+  $ ./autokft.sh sys_write ./ 1 # this will be triggered by external actions
+
 please learn more the kd usage from Documentation/kft.txt
 
-there is all, be happy to play with it.
+that is all, be happy to play with it.
diff --git a/scripts/autokft.sh b/scripts/autokft.sh
new file mode 100755
index 0000000..446374d
--- /dev/null
+++ b/scripts/autokft.sh
@@ -0,0 +1,86 @@
+#!/bin/bash
+# autokft.sh -- trace an indicated function automatically
+# author: falcon <wuzhangjin@gmail.com>
+# update: 2009-08-06
+# usage:
+#      $ ./autokft.sh [function_name] [script_path] [1|0]
+# if you have copied the tools: scripts/{kd,sym2addr,addr2sym} to current directory, try this:
+# i.e. $ ./autokft.sh sys_write ./ 1
+
+function error_report
+{
+	echo "usage: "
+	echo "    $ ./autokft.sh [function_name] [script_path] [1|0]"
+	echo "    note: you can copy the tools: scripts/{kd,sym2addr,addr2sym} from the linux kernel source code to current directory, and then try this"
+	echo "    $ ./autokft.sh sys_write ./    # trigger it ourselves"
+	echo "    or"
+	echo "    $ ./autokft.sh sys_write ./ 1  # trigger by external actions"
+	exit
+}
+
+# get the function need to tace from user
+[ -z "$1" ] && echo "please input the function need to be traced" && error_report
+
+trac_func=$1
+
+# get the path of the path of the tools: scripts/{sym2addr,addr2sym,kd}
+script_path=  #/path/to/kernel/usr/src/scripts/
+
+[ -n "$2" ] && script_path=$2
+
+if [ -z "$script_path" ]; then
+	echo "please configure the path of scripts as script_path" && error_report
+fi
+
+# start it manually or automatically
+auto=0		# if you want to trace it by external trigger, change it to 1
+
+[ -n "$3" ] && auto=$3
+
+# generate the latest system.map from /proc/kallsyms 
+if [ ! -f system.map ]; then
+	cat /proc/kallsyms > system.map
+fi
+
+# generate a default configuration file for kft
+cat <<EOF > config.sym
+new
+begin
+	trigger start entry $trac_func
+	trigger stop exit $trac_func
+end
+EOF
+
+# convert the symbols to address via the system.map
+$script_path/sym2addr config.sym system.map > config.addr
+
+# config kft
+cat config.addr > /proc/kft
+
+# prime it
+echo prime > /proc/kft
+
+sleep 1 
+
+# start it
+
+if [ "$auto" -eq 1 ];then
+	grep -q "not complete" /proc/kft
+	while [ $? -eq 0 ]
+	do
+		echo "please do something in the other console or terminal to trigger me"
+		sleep 1
+	done
+else
+	echo start > /proc/kft
+fi
+sleep 1
+
+# get the data
+cat /proc/kft_data > log.addr
+
+# convert the address to symbols
+$script_path/addr2sym < log.addr -m system.map > log.sym
+
+# generate a readable log
+$script_path/kd -c -l -i log.sym
-- 
1.9.1

